python/ql/lib/semmle/python/dataflow/new/internal/DataFlowPrivate.qll

```diff
@@ -376,6 +367,34 @@ module EssaFlow {
   predicate defToFirstUse(EssaVariable var, NameNode nodeTo) {
     AdjacentUses::firstUse(var.getDefinition(), nodeTo)
   }
+
+  predicate useUseFlowStep(Node nodeFrom, Node nodeTo) {
+    // First use after definition
+    //   `y = 42`
+    //   `x = f(y)`
+    //   nodeFrom is `y` on first line, essa var
+    //   nodeTo is `y` on second line, cfg node
+    defToFirstUse(nodeFrom.asVar(), nodeTo.asCfgNode())
+    or
+    // Next use after use
+    //   `x = f(y)`
+    //   `z = y + 1`
+    //   nodeFrom is 'y' on first line, cfg node
+    //   nodeTo is `y` on second line, cfg node
+    useToNextUse(nodeFrom.asCfgNode(), nodeTo.asCfgNode())
+  }
+
+  predicate localFlowStep(Node nodeFrom, Node nodeTo) {
+    IncludePostUpdateFlow<PhaseDependentFlow<definitionFlowStep/2>::step/2>::step(nodeFrom, nodeTo)
+    or
+    IncludePostUpdateFlow<PhaseDependentFlow<expressionFlowStep/2>::step/2>::step(nodeFrom, nodeTo)
+    or
+    // Use-use flow can generate self loops. We want to filter steps from `n` to `n`
+    // after we have included steps from `[post] n` to `n`, so after
+    // `IncludePostUpdateFlow` has ben applied.
```

Comment:
The term "self loops" was confusing to me. Can we use the standard wording from the consistency tests, that is `Node steps to itself`?

```suggestion
    // Blindly applying use-use flow can result in a node that steps to itself, for
    // example in while-loops. To uphold dataflow consistency checks, we don't want
    // that. However, we do want to allow `[post] n` to `n` (to handle while loops), so
    // we should only do the filtering after `IncludePostUpdateFlow` has ben applied.
```